### PR TITLE
Optimize GC.AllocateUninitializedArray and use it in StringBuilder

### PR DESCRIPTION
**Pull request by adamsitnik**
I wanted to use `GC.AllocateUninitializedArray` in `StringBuilder`, but it was initially too slow. Calling it for small buffers was causing quite noticeable performance degradation. 

I've tuned it up to ensure that it does not slow down the `StringBuilder` in "unlucky path" (small arrays) and does improve the perf in "lucky path" (big arrays). It should make this API more profitable to use in other places in the future.

Changes:
* remove one branch by changing the `int` to `uint` (`if (length < 0)`). This changes the behavior of this internal API for `length < 0` - previously the caller would get `IndexOutOfRangeException`. Since this is an internal API, I hope it's OK.
* increase the threshold from 256 to 2048 bytes - please see the results below. 
* enforce inlining of `GC.AllocateUninitializedArray`, move the expensive native call to separate method to not increase the size too much.

Micro benchmarks for the GC API:

```cs
[GenericTypeArguments(typeof(byte))]
[GenericTypeArguments(typeof(char))]
[GenericTypeArguments(typeof(object))]
public class Perf_GC<T>
{
    private readonly Func<int, T[]> _allocateUninitializedArrayDelegate = CreateDelegate<int>(typeof(GC), "AllocateUninitializedArray");
    private readonly Func<int, T[]> _allocateArrayDelegate = CreateDelegate<int>(typeof(Mimic), "AllocateArray");

    [Params(256, 256 * 2, 256 * 3, 256 * 4, 256 * 6, 256 * 8)]
    public int Length;

    [Benchmark]
    public T[] AllocateUninitializedArray() => _allocateUninitializedArrayDelegate(Length);

    [Benchmark]
    public T[] AllocateArray() => _allocateArrayDelegate(Length); // using delegate for apples to apples comparison

    private static Func<N, T[]> CreateDelegate<N>(Type type, string methodName)
    {
        // this method is not a part of .NET Standard so we need to use reflection
        var method = type
            .GetMethod(methodName, BindingFlags.NonPublic | BindingFlags.Static)
            .MakeGenericMethod(typeof(T));

        return method != null ? (Func<N, T[]>)method.CreateDelegate(typeof(Func<N, T[]>)) : null;
    }
}

public static class Mimic
{
    internal static T[] AllocateArray<T>(int size) => new T[size];
}
```

I've simplified the default BDN output to make it easier to compare the results. In the table below the "Before" is the execution time for `GC.AllocateUninitializedArray` before my changes, in the "After" are with my changes. The `new T[]` contains the time for calling new operator (to have some base comparison)

|            Type | Length |    Before |     After |   new T[] |
|---------------- |------- |----------:|----------:|----------:|
|            Byte |    256 |  78.63 ns |  18.31 ns |  18.17 ns |
|            Char |    256 |  79.33 ns |  31.95 ns |  31.66 ns |
|          Object |    256 | 113.34 ns | 113.34 ns | 113.03 ns |
|                 |        |           |           |           |
|            Byte |    512 |  79.37 ns |  31.38 ns |  31.75 ns |
|            Char |    512 |  87.60 ns |  58.12 ns |  57.71 ns |
|          Object |    512 | 229.02 ns | 229.30 ns | 227.78 ns |
|                 |        |           |           |           |
|            Byte |    768 |  83.24 ns |  45.51 ns |  45.85 ns |
|            Char |    768 |  95.92 ns |  85.20 ns |  84.34 ns |
|          Object |    768 | 353.66 ns | 347.39 ns | 349.48 ns |
|                 |        |           |           |           |
|            Byte |   1024 |  85.99 ns |  58.31 ns |  57.58 ns |
|            Char |   1024 |  99.46 ns | 100.62 ns | 112.01 ns |
|          Object |   1024 | 457.07 ns | 455.94 ns | 457.47 ns |
|                 |        |           |           |           |
|            Byte |   1536 |  92.40 ns |  84.84 ns |  84.44 ns |
|            Char |   1536 | 111.75 ns | 112.97 ns | 168.02 ns |
|          Object |   1536 | 653.64 ns | 649.47 ns | 643.37 ns |
|                 |        |           |           |           |
|            Byte |   2048 | 100.61 ns | 101.04 ns | 111.81 ns |
|            Char |   2048 | 126.52 ns | 125.31 ns | 226.94 ns |
|          Object |   2048 | 830.92 ns | 838.90 ns | 836.48 ns |

### CODE REVIEW

**Comment by adamsitnik** [comment]
The `StringBuilder` benchmarks (that show a difference, I did not include the benchmarks for which the results have not changed):

```
public class Perf_StringBuilder
{
    const int LOHAllocatedStringSize = 100_000;

    private string _stringLOH = new string('a', LOHAllocatedStringSize);
    private string _string100 = new string('a', 100);

    [Benchmark]
    [Arguments(100)]
    [Arguments(LOHAllocatedStringSize)]
    public StringBuilder ctor_string(int length) => new StringBuilder(length == 100 ? _string100 : _stringLOH);

    [Benchmark]
    [Arguments(100)]
    [Arguments(LOHAllocatedStringSize)]
    public StringBuilder ctor_capacity(int length) => new StringBuilder(length);

    [Benchmark]
    [Arguments(100)]
    [Arguments(LOHAllocatedStringSize)]
    public StringBuilder Append_Char_Capacity(int length)
    {
        StringBuilder builder = new StringBuilder(length);

        for (int i = 0; i < length; i++)
        {
            builder.Append('a');
        }

        return builder;
    }

    [Benchmark]
    [Arguments(1)]
    [Arguments(1_000)]
    public StringBuilder Append_Strings(int repeat)
    {
        StringBuilder builder = new StringBuilder();

        // strings are not sorted by length to mimic real input
        for (int i = 0; i < repeat; i++)
        {
            builder.Append("12345");
            builder.Append("1234567890abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMN");
            builder.Append("1234567890abcdefghijklmnopqrstuvwxy");
            builder.Append("1234567890");
            builder.Append("1234567890abcdefghijklmnopqrstuvwxyzABCDEFGHI");
            builder.Append("1234567890abcde");
            builder.Append("1234567890abcdefghijklmnopqrstuvwxyzABCD");
            builder.Append("1234567890abcdefghijklmnopqrst");
            builder.Append("1234567890abcdefghij");
            builder.Append("1234567890abcdefghijklmno");
        }

        return builder;
    }
}
```

|                    Method |                         Toolchain | repeat | length |          Mean | Ratio | Allocated |
|-------------------------- |---------------------------------- |------- |------- |--------------:|------:|----------:|
|            Append_Strings |      \Tests\Core_Root\CoreRun.exe |   1000 |      ? | 119,591.06 ns |  0.97 |  559264 B |
|            Append_Strings | \sync_bound\Core_Root\CoreRun.exe |   1000 |      ? | 122,728.26 ns |  1.00 |  559264 B |
|                           |                                   |        |        |               |       |           |
|               ctor_string |      \Tests\Core_Root\CoreRun.exe |      ? | 100000 |  16,553.47 ns |  0.72 |  200072 B |
|               ctor_string | \sync_bound\Core_Root\CoreRun.exe |      ? | 100000 |  23,078.37 ns |  1.00 |  200072 B |
|                           |                                   |        |        |               |       |           |
|             ctor_capacity |      \Tests\Core_Root\CoreRun.exe |      ? | 100000 |   6,276.58 ns |  0.48 |  200072 B |
|             ctor_capacity | \sync_bound\Core_Root\CoreRun.exe |      ? | 100000 |  13,019.77 ns |  1.00 |  200072 B |
|                           |                                   |        |        |               |       |           |
|      Append_Char_Capacity |      \Tests\Core_Root\CoreRun.exe |      ? | 100000 | 219,339.17 ns |  0.97 |  200072 B |
|      Append_Char_Capacity | \sync_bound\Core_Root\CoreRun.exe |      ? | 100000 | 226,150.72 ns |  1.00 |  200072 B |

**Comment by adamsitnik** [comment]
BTW the native call path could benefit a lot from https://github.com/dotnet/coreclr/issues/5329, I am going to add a comment there

**Comment by stephentoub** [comment]
@adamsitnik, I don't understand the StringBuilder benchmarks: these all show them getting slower.  What am I missing?

**Comment by adamsitnik** [comment]
> I don't understand the StringBuilder benchmarks: these all show them getting slower.  What am I missing?

I am sorry when I was hand-editing the path to CoreRun to replace it with "before" and "after" I've introduced a bug ;p

File: src/System.Private.CoreLib/src/System/GC.cs [hunk 1]
```diff
@@ -651,32 +649,29 @@ internal static void UnregisterMemoryLoadChangeNotification(Action notification)
...
 
-        // Skips zero-initialization of the array if possible. If T contains object references,
-        // the array is always zero-initialized.
-        internal static T[] AllocateUninitializedArray<T>(int length)
+        /// <summary>
+        /// Skips zero-initialization of the array if possible. If T contains object references,
+        /// the array is always zero-initialized.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)] // forced to ensure no perf drop for small memory buffers (hot path)
+        internal static T[] AllocateUninitializedArray<T>(uint length)
```
Comment (jkotas) [inline src/System.Private.CoreLib/src/System/GC.cs]: This should stay as `int`. `int` is in approved shape of the public API (https://github.com/dotnet/corefx/issues/31787). We just have not made this API public yet, and we do not want to be forking the API shape for internal use.

File: src/System.Private.CoreLib/src/System/GC.cs [hunk 2]
```diff
@@ -651,32 +649,29 @@ internal static void UnregisterMemoryLoadChangeNotification(Action notification)
...
                 return new T[length];
             }
-#endif
-            return (T[])AllocateNewArray(typeof(T[]).TypeHandle.Value, length, zeroingOptional: true);
+
+            // kept outside of the small arrays hot path to have inlining without big size growth
+            return AllocateUninitializedArray_NativeCall<T>(length);
         }
+
+        private static T[] AllocateUninitializedArray_NativeCall<T>(uint length)
```
Comment (jkotas) [inline src/System.Private.CoreLib/src/System/GC.cs]: Why did you rename this? `_NativeCall` suffix is not used anywhere else in CoreLib for FCalls. The original name looked just fine to me.

File: src/System.Private.CoreLib/src/System/GC.cs [hunk 3]
```diff
@@ -651,32 +649,29 @@ internal static void UnregisterMemoryLoadChangeNotification(Action notification)
...
+        internal static T[] AllocateUninitializedArray<T>(uint length)
         {
             if (RuntimeHelpers.IsReferenceOrContainsReferences<T>())
             {
                 return new T[length];
             }
 
-            if (length < 0)
-                ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.lengths, 0, ExceptionResource.ArgumentOutOfRange_NeedNonNegNum);
-#if DEBUG
```
Comment (jkotas) [inline src/System.Private.CoreLib/src/System/GC.cs]: Why did you remove `#if DEBUG`? It was here to improve the code coverage and I believe that it was proven to be useful to find number of bugs earlier.

File: src/System.Private.CoreLib/src/System/GC.cs (hunk 1 above)
Comment (VSadov) [reply src/System.Private.CoreLib/src/System/GC.cs]: Yes, the argument should be int.

**Comment by VSadov** [comment]
Another interesting case would be if an uninitialized allocation is mixed with ordinary allocation. Allocating one object after each array should do it.
I’d expect the breakeven threshold to be higher for such case.

**Comment by jkotas** [comment]
@GrabYourPitchforks @bartonjs Thoughts about this from the security point of view?

**Comment by GrabYourPitchforks** [comment]
I can't think of anything offhand. As long as there's no way for a caller to observe the uninitialized data, you should be good. For instance, `StringBuilder.set_Length` should continue to zero-fill the padding to overwrite any previously existing data. `StringBuilder.GetChunks` should continue to return an enumerator whose `ReadOnlyMemory<char>` elements are properly sliced. (If a caller then uses `MemoryMarshal` or unsafe code to get at the rest of the data, they're taking responsibility for their own actions.)

**Comment by adamsitnik** [comment]
> Another interesting case would be if an uninitialized allocation is mixed with ordinary allocation. Allocating one object after each array should do it.
I’d expect the breakeven threshold to be higher for such case.

I've modified the benchmark to include single object allocation before allocating the array and the threshold grew to 7kb.

**Comment by VSadov** [comment]
7k - is that with `int` parameter or `uint`?

**Comment by bartonjs** [comment]
In general I'm not a fan of "malloc vs calloc"; but if we have a plethora of tests then we can feel pretty confident that we're never acting on the uninitialized data (either StringBuilder internally or it being returned to callers).

File: src/System.Private.CoreLib/src/System/GC.cs [hunk 4]
```diff
@@ -661,22 +662,22 @@ internal static T[] AllocateUninitializedArray<T>(int length)
...
-            // otherwise small arrays are allocated using `new[]` as that is generally faster.
-            //
-            // The threshold was derived from various simulations.
-            // As it turned out the threshold depends on overal pattern of all allocations and is typically in 200-300 byte range.
-            // The gradient around the number is shallow (there is no perf cliff) and the exact value of the threshold does not matter a lot.
-            // So it is 256 bytes including array header.
-            if (Unsafe.SizeOf<T>() * length < 256 - 3 * IntPtr.Size)
+            }
+
+#if !DEBUG  // for debug builds we always want to call AllocateNewArray to detect AllocateNewArray bugs
```
Comment (bartonjs) [inline src/System.Private.CoreLib/src/System/GC.cs]: The trailing comment aligned to be a statement comment feels weird to me.  Would you mind inserting a linebreak between the #if and the comment?

Comment (jkotas) [reply src/System.Private.CoreLib/src/System/GC.cs]: I would but the comment in front of the ifdef. The comments are typically placed before the code line that they apply to.

File: src/System.Private.CoreLib/src/System/GC.cs [hunk 5]
```diff
@@ -661,22 +662,22 @@ internal static T[] AllocateUninitializedArray<T>(int length)
             }
 
             if (length < 0)
```
Comment (jkotas) [inline src/System.Private.CoreLib/src/System/GC.cs]: This should be in the unmanaged FCall - no need to force-inline the error handling everywhere. Also, we should make the exception same as the exception thrown by `new T[length]` for negative length.

File: src/System.Private.CoreLib/src/System/GC.cs [hunk 6]
```diff
@@ -661,22 +662,22 @@ internal static T[] AllocateUninitializedArray<T>(int length)
...
             {
                 return new T[length];
             }
 #endif
-            return (T[])AllocateNewArray(typeof(T[]).TypeHandle.Value, length, zeroingOptional: true);
+            // kept outside of the small arrays hot path to have inlining without big size growth
+            return AllocateNewUninitializedArray(length);
+
+            T[] AllocateNewUninitializedArray(int length)
+                => (T[])AllocateNewArray(typeof(T[]).TypeHandle.Value, length, zeroingOptional: true);
```
Comment (jkotas) [inline src/System.Private.CoreLib/src/System/GC.cs]: Use `Unsafe.As` here to make this faster / smaller?

File: src/System.Private.CoreLib/src/System/GC.cs (hunk 5 above)
Comment (VSadov) [reply src/System.Private.CoreLib/src/System/GC.cs]: FCall passes everything to the same helper as for normal arrays (just with a noinit flag). 
It all was refactored to go through `AllocateSzArray` internally, which has the checks for sizes, dimensions etc...

The negative length check is not necessary here. I do not remember why it was added. Do we care from where the exception would be thrown? (it think not)

Comment (jkotas) [reply src/System.Private.CoreLib/src/System/GC.cs]: Right, we do not care - as long as the exception is consistent.

File: src/System.Private.CoreLib/src/System/GC.cs [hunk 7]
```diff
@@ -661,22 +662,22 @@ internal static T[] AllocateUninitializedArray<T>(int length)
...
+            if (length < 2048 / Unsafe.SizeOf<T>())
             {
                 return new T[length];
             }
 #endif
-            return (T[])AllocateNewArray(typeof(T[]).TypeHandle.Value, length, zeroingOptional: true);
+            // kept outside of the small arrays hot path to have inlining without big size growth
+            return AllocateNewUninitializedArray(length);
+
+            T[] AllocateNewUninitializedArray(int length)
```
Comment (jkotas) [inline src/System.Private.CoreLib/src/System/GC.cs]: Maybe add a comment here with the link to the issue about intrinsic `TypeHandle.Value` so that this split can be undone once that issue is fixed?

**Comment by adamsitnik** [comment]
> 7k - is that with int parameter or uint?

int

File: src/System.Private.CoreLib/src/System/GC.cs (hunk 7 above)
Comment (adamsitnik) [reply src/System.Private.CoreLib/src/System/GC.cs]: good idea, done

**Comment by adamsitnik** [comment]
> once the CI is green

I've removed following length check from `AllocateNewArray`:

https://github.com/dotnet/coreclr/blob/310f71592b1373297a98ae268f795ca6dedd62e6/src/vm/comutilnative.cpp#L1134

Because `AllocateSzArray` is responsible for throwing the exception for negative length:

https://github.com/dotnet/coreclr/blob/ab2f9caaf35c96d029b96aa171ee65d04253cf7c/src/vm/gchelpers.cpp#L446-L447
